lib/host/wasi/inode-win.cpp

```
@@ -330,6 +330,26 @@ SymlinkPriviledgeHolder SymlinkPriviledgeHolder::Holder;
 
 } // namespace
 
+namespace {
+static winapi::HANDLE_ getWindowsHandle(int32_t Fd, winapi::DWORD_ StdHandleId,
+                                        std::string_view Name) {
+  if (Fd < 0) {
+    return winapi::GetStdHandle(StdHandleId);
+  }
+
+  auto raw = _get_osfhandle(Fd);
+  auto Handle = reinterpret_cast<winapi::HANDLE_>(raw);
+
+  if (Handle != winapi::INVALID_HANDLE_VALUE_ && Handle != nullptr) {
+    return Handle;
+  }
+  spdlog::warn("    Failed to get OS handle for custom {} FD: {}. "
+               "Falling back to default.",
+               Name, Fd);
```

I would like to discuss this behavior: do we really want users to receive a default FD when the provided FD is invalid?  
If so, we must document this behavior so that people understand what is happening when the given FD is broken or invalid. Otherwise, we should report a fatal error to indicate that this is unacceptable.

cc @q82419 @ibmibmibm

---

@hydai I agree we should either document this fallback or make it a hard failure. Open to feedback on which direction makes more sense here.